core/src/main/scala/neotypes/model.scala

```diff
@@ -0,0 +1,198 @@
+package neotypes
+package model
+
+import mappers.{ParameterMapper, ResultMapper}
+
+import org.neo4j.driver.types.{IsoDuration => NeoDuration, Point => NeoPoint}
+
+import java.time.{LocalDate => JDate, LocalDateTime => JDateTime, LocalTime => JTime, OffsetTime => JZTime, ZonedDateTime => JZDateTime}
+import java.util.{Map => JMap}
+import scala.collection.immutable.ArraySeq
+import scala.jdk.CollectionConverters._
+import scala.util.control.NoStackTrace
+
+/** Safe wrapper over a Neo4j query parameter. */
+final class QueryParam private[neotypes] (private[neotypes] val underlying: AnyRef) extends AnyVal
+object QueryParam {
+  def apply[A](scalaValue: A)(implicit mapper: ParameterMapper[A]): QueryParam =
+    mapper.toQueryParam(scalaValue)
+
+  private[neotypes] def toJavaMap(map: Map[String, QueryParam]): JMap[String, Object] =
+    map.map {
+      case (key, value) =>
+        key -> value.underlying
+    }.asJava
+}
+
+/** Data types supported by Neo4j. */
+object types {
+  /** Parent type of all Neo4j types. */
+  sealed trait NeoType extends Product with Serializable
```

Comment:
NIt: I prefer using `sealed abstract class` for ADTs. It enforces single-inheritance, is binary-compatible with Scala 3 enum, and theoretically has slightly better performance :)